### ISSUE TITLE
Per-sphere mincollapsecount and wild matches

### ISSUE THREAD

**Issue by TF3RDL** [jakubg1/OpenSMCE#65]
Some SM games like Tumblebugs have a wild ball thing that detonates even if you place a wild sphere at the first or last part of the train and what meets on one side is a colored single chain and nothing else on the other, which is not a three or more ball match. Things like this can be replicated by implementing the setting to set mincollapsecount for each sphere types (in the case, wild spheres have mincollapsecount of 1) and if the aforementioned property is not defined, it will be set to three by default

I also like to add the `match_type` attribute for spheres; `normal` is the same as before, `wild` means spheres can match to anything and it can destroy both sides even though it won't make a match if you shoot a different colored sphere at different color adjacent to wild sphere similar to in Luxor QFTA, and `rock` is the same as setting `matches` attribute to `[-4, 1, 2, 3, 4, 5, 6, 7]` or something like that and not setting other spheres to add -4 ID to its sphere matches list

**Comment by jakubg1** [comment]
Minimum match count is planned for the future. I'm not sure when exactly, but when added, I'll add a possibility of overridding it per sphere type.
The `match_type` field is already reserved for something different: depending on the value, I plan to add a feature of just "gluing" the sphere to the train without appending upon a successful match. This will take some time, it's not easy.

The `match_type` attribute you described is planned to be generalized in some form of splitting the `matches` field into categories as: considering a shot as successful or missed, considering a sphere to be destroyed upon match, considering a sphere to be magnetizable, and maybe more. This also will involve adding sphere tags.

The color replacer behavior will not be implemented for now because it seems to be a side effect of your solution. However, it is planned to make it more flexible via sphere selectors (this will also remove the distinction between color clouds and color replacers).

**Comment by TF3RDL** [comment]
Yeah. Maybe also add the per-sphere `spawnstreak`, which specifies the average clump size for particular sphere types; setting it to zero means the specified sphere type is single-group only (in the case of wild balls appearing on the path, the wild sphere's `spawnstreak` is 0, which means only single groups for wild balls appearing) and vice versa

As for the proposed `match_type` parameter, the wild spheres in Luxor: Amun Rising and Sparkle 2 does make a match even when shooting different color at the wild sphere where other side adjacent to it is a different color, so I proposed a `wild2` value so it matches like ones from Luxor AR but without two bugs (powerups not spawning on 3x combos when this sphere type is present and matches after line merges)

**Comment by jakubg1** [comment]
As for the per-color spawnstreak, I'm planning a new system of spawning spheres. This is on a very long to-do list, so it may take some time.

The wild sphere behavior in some games is really ambiguous, so I'll leave it till later. There are much more important things to do.

**Comment by TF3RDL** [comment]
The `stone` type has already been implemented in available value on `type` in spheres. I propose the sphere behavior for `wild` type (can destroy two sides of groups of different sphere types including obstacle spheres adjacent to wild ball, attracts to anything, and it doesn't match if the player puts a different colored sphere on either side whereas the other side has a different color or being attracted to a different color where again other side have different color)

**Comment by jakubg1** [comment]
I will consider this in the future. There are more serious problems with the engine which need to be resolved right now.

**Comment by TF3RDL** [comment]
Since sphere types with different behaviors than normal balls changes the game difficulty a _lot_ more than just adding more colors, the per-sphere score when destroyed would be nice (like sphere types that are difficult to deal with than normal worths more points than regular one and vice versa), and oh yeah, I forgot the magnetization table thing that determines the spheres should be attracted to another one if it is in the table

**Comment by jakubg1** [comment]
Score events are planned as one of the Beta 4.8.0's features. Magnetization table - thought about that, but I'll see what can I do later.

**Comment by TF3RDL** [comment]
Lately I have played [AudioSurf](https://www.youtube.com/watch?v=Zy7D8dVFimg) so, the [black/white blocks](https://audiosurf.fandom.com/wiki/White_blocks) mechanic (which is unmatchable but can be removed and gives bonus points when goes to the bottom) gave me an idea of an `edible` sphere type, which is where it gets removed once it hits a scarab/sphere pusher and perhaps an "inverse" variant where destroying spheres on front of these will get removed and gives off bonus points

However, both of them could have edge cases; what if the `edible` sphere spawns on the front of a sphere pusher and/or the `inverse-edible` one appears on the end of the sphere train? and how about the spheres with `edible` type in a Zuma-style contiguous sphere chain?

**Comment by TF3RDL** [comment]
> While the idea is pretty cool, it is pretty niche and I'm not sure if there are any games in an advanced development stage which would use this mechanic, so it will unfortunately wind up with a very low priority. Maybe at some point though!

The sphere with match type of `inverse-edible` I've described on other marble popper games I can think of are rock balls on Woka Woka where you must destroy all spheres in front of these balls in order to get rid of them. But I haven't aware of the other sphere-matching games with a sphere type of `edible` match type where you must destroy all spheres behind these in order to get "eaten" by a sphere pusher so, let me know if there is one with `edible` match type in other SM games

BTW if you're adding `wild` and `wild2` match types (which can hopefully making adding new sphere colors/types less tedious in the presence of wild balls), I think the "exclude wild matches" property for some sphere types should be added if your sphere type should not match with and be destroyed by wild balls nearby matching with other colors

**Comment by jakubg1** [comment]
Naming types `wild` and `wild2` is a very bad practice. I would use additional words to describe it as accurately as possible.
At some point, I do plan to split the sphere lists, so they would be separate for magnetization, matching, etc. Just not right now.

**Comment by TF3RDL** [comment]
> Naming types `wild` and `wild2` is a very bad practice. I would use additional words to describe it as accurately as possible.

I admit that my naming conventions are bad, so what about `pseudo-wild` and `wild` type?
> At some point, I do plan to split the sphere lists, so they would be separate for magnetization, matching, etc. Just not right now.

This should allow for a sphere type where it matches like normal but it never attract at all
